### ISSUE TITLE
Differences between lumaku and cornerfarmer implementations

### ISSUE THREAD

**Issue by ShantanuNair** [lumaku/ctc-segmentation#24]
First of all marvelous piece of work done here! Thanks @lumaku , your continued participation in ASR projects as well is invaluable!!

I had a query regarding differences between the implementation and features exposed by this repo and the repo at [cornerfarmers repo](https://github.com/cornerfarmer/ctc_segmentation).

Are there any differences wrt implementation? What about
- Performance? GPU computability - I understand the algo is not meant for the GPU and instead better on a strong single core CPU unit, but mentions of using RNN instead confuse me. Perhaps that's only for getting the logits? This part confused me a little, as well as trying to search for such a suitable pretrained RNN+ctc based character outputting STT implementation.
- Working with longer audio files
- Any other interfaces exposed or features provided

I understand that this repo is based off on the cornerfarmer one, as that is the code for the paper ([DOI: 10.1007/978-3-030-60276-5_27](https://doi.org/10.1007/978-3-030-60276-5_27)), but would like to ask the author @lumaku if there are any insights to be gained here.

Of particular use would be for my use case of force aligning long form audio using ctc-segmentation using ASR-generated transcripts. Any insights regarding this would be appreciated as well, otherwise I can create a new topic if that is more acceptable! 

Thanks again @lumaku !

**Comment by lumaku** [comment]
This repository contains a python package that is adapted for and integrated into several toolkits: Espnet 2, Speechbrain and Nemo. You can download one of these toolkits and start aligning.

The cornerfarmer repository archives the code of our paper - it is organized in the form of a recipe (as in, a kaldi recipe). The original recipe-style is still available in the ESPnet 1 toolkit. However, the toolkits mentioned above provide a direct Python command line interface that makes it more user-friendly.

About working on longer audio files: If the text fits to the audio, then alignment should not take longer than a second, even on a few-hours long audio file. If the text is garbage, it might take up to a few seconds until it reports a failed alignment (which is not the case here). This process is parallelizable, and basically limited by the RAM usage.

Usually ASR inference consumes most resources. Therefore, my recommendations about RNN/Transformer are only regarding the inference network to obtain the CTC activations: Today, RNNs are a bit outdated and Transformer-based networks are the standard. Without and additional tricks, Transformers are extremely resource-hungry for longer sequences. You don't need to train an RNN for your long audio files, it is possible to split the audio files into smaller parts. There a point about this in the FAQ. I recommend that.

In comparison to the archived implementation from our paper, there is an additional feature that improves alignment speed and reduces memory consumption: Directly aligning from tokens. However, this is best used with a character-based ASR model, e.g. `["A", "B", "C", ...]`; it may reduce accuracy if used with a model with many and longer tokens. Most ASR toolkits provide those character-based models. This also helps for longer audio files.

I think that aligning your own inferred text is an interesting idea. Maybe the per-utterance confidence score is useful. You would still need to find a good way to separate utterances in a meaningful way, as most ASR models don't have punctuation and sentence endings in their token dictionaries. Or, you could just align the words....

If you plan to use the same model for aligning and inference, you could directly store the CTC logits during inference for the alignment afterwards. The CTC segmentation interfaces of Speechbrain and ESPnet 2 may give you inspiration of how to implement this (see the `get_lpz()` function). Running the alignment in parallel would not use that much more resources.
